enterprise/pulsar.py

```diff
@@ -135,12 +135,20 @@ def sort_data(self):
             self._isort = slice(None, None, None)
             self._iisort = slice(None, None, None)
 
-    def filter_data(self, start_time=None, end_time=None):
-        """Filter data to create a time-slice of overall dataset."""
-        if start_time is None and end_time is None:
-            mask = np.ones(self._toas.shape, dtype=bool)
-        else:
-            mask = np.logical_and(self._toas >= start_time * 86400, self._toas <= end_time * 86400)
+    def filter_data(self, mask=None, start_time=None, end_time=None):
+        """
+        Filters the dataset to create a time-slice based on a custom mask or time range.
+
+        Parameters:
+            mask (array-like, optional): Boolean array specifying which data to keep.
+                                         If None, `start_time` and `end_time` are used. Default is None.
+            start_time (float, optional): Start time (MJD) for filtering. Ignored if `mask` is provided. Default None.
+            end_time (float, optional): End time (MJD) for filtering. Ignored if `mask` is provided. Default None.
+        """
+
+        start_time = start_time * 86400 if start_time is not None else np.min(self._toas)
+        end_time = end_time * 86400 if end_time is not None else np.max(self._toas)
+        mask = mask if mask is not None else np.logical_and(self._toas >= start_time, self._toas <= end_time)
```

Comment:
Do we want to combine, with a logical and, the mask that was given, and the mask from start_time and end_time?

Reply:
Yes that's a good idea, I updated this using your suggestion below!